components/pd_client/src/client_v2.rs

```diff
@@ -0,0 +1,1339 @@
+    ) -> PdFuture<()> {
+        let timer = Instant::now();
+        let mut req = pdpb::UpdateServiceGcSafePointRequest::default();
+        req.set_header(self.header());
+        req.set_service_id(name.into());
+        req.set_ttl(ttl.as_secs() as _);
+        req.set_safe_point(safe_point.into_inner());
+
+        let mut raw_client = self.raw_client.clone();
+        Box::pin(async move {
+            raw_client.wait_for_ready().await?;
+            let resp = raw_client
+                .stub()
+                .update_service_gc_safe_point_async_opt(
+                    &req,
+                    raw_client
+                        .call_option()
+                        .timeout(Duration::from_secs(REQUEST_TIMEOUT)),
+                )
+                .unwrap_or_else(|e| {
+                    panic!(
+                        "fail to request PD {} err {:?}",
+                        "update_service_safe_point", e
+                    );
+                })
+                .await?;
+            PD_REQUEST_HISTOGRAM_VEC
+                .with_label_values(&["update_service_safe_point"])
+                .observe(duration_to_sec(timer.saturating_elapsed()));
+            check_resp_header(resp.get_header())?;
+            Ok(())
+        })
+    }
+
+    fn report_min_resolved_ts(&mut self, store_id: u64, min_resolved_ts: u64) -> PdFuture<()> {
+        let timer = Instant::now();
+
+        let mut req = pdpb::ReportMinResolvedTsRequest::default();
+        req.set_header(self.header());
+        req.set_store_id(store_id);
+        req.set_min_resolved_ts(min_resolved_ts);
+
+        let mut raw_client = self.raw_client.clone();
+        Box::pin(async move {
+            raw_client.wait_for_ready().await?;
+            let resp = raw_client
+                .stub()
+                .report_min_resolved_ts_async_opt(
+                    &req,
+                    raw_client
+                        .call_option()
+                        .timeout(Duration::from_secs(REQUEST_TIMEOUT)),
+                )
+                .unwrap_or_else(|e| {
+                    panic!("fail to request PD {} err {:?}", "min_resolved_ts", e);
+                })
+                .await?;
+            PD_REQUEST_HISTOGRAM_VEC
+                .with_label_values(&["min_resolved_ts"])
+                .observe(duration_to_sec(timer.saturating_elapsed()));
+            check_resp_header(resp.get_header())?;
+            Ok(())
+        })
+    }
+}
+
+async fn reconnect_loop(
+    mut client: CachedRawClient,
+    on_reconnect: CallbackHolder,
+    mut reconnect: mpsc::UnboundedReceiver<u64>,
+) {
+    let interval = Duration::from_secs(REQUEST_TIMEOUT);
+    loop {
+        while !client.channel().wait_for_connected(interval).await {
+            // wait_for_connected will attempt to connect internally. When it fails we
+            // should force reconnect.
+            if let Err(e) = client.reconnect(true, &on_reconnect).await {
```

Comment:
If leader is still the same and member is still the same, there is no reason to build a new client.